python/ray/data/read_api.py

```diff
@@ -1053,11 +1063,73 @@ def _get_metadata(table: Union["pyarrow.Table", "pandas.DataFrame"]) -> BlockMet
 
 
 def _prepare_read(
-    ds: Datasource, ctx: DatasetContext, parallelism: int, kwargs: dict
+    ds: Datasource,
+    ctx: DatasetContext,
+    cur_pg: Optional[PlacementGroup],
+    parallelism: int,
+    kwargs: dict,
 ) -> List[ReadTask]:
     kwargs = _unwrap_arrow_serialization_workaround(kwargs)
     DatasetContext._set_current(ctx)
-    return ds.prepare_read(parallelism, **kwargs)
+    reader = ds.create_reader(**kwargs)
+    parallelism = _autodetect_parallelism(parallelism, cur_pg, reader)
+    return reader.prepare_read(parallelism)
+
+
+def _autodetect_parallelism(
+    parallelism: int, cur_pg: Optional[PlacementGroup], reader=None
+) -> int:
+    # Autodetect parallelism requested. The heuristic here are that we should try
+    # to create as many blocks needed to saturate available resources, and also keep
+    # block sizes below the target memory size, but no more. Creating too many
+    # blocks is inefficient.
+    if parallelism < 0:
```

Comment:
The auto-detected parallelism will be truncated to the number of files in the datasource `prepare_read()` functions, which I think should be sufficient. Although we should note that this may result the target block size to be exceeded, but that will eventually be solved by dynamic block splitting.

Reply:
I think it's better to centralize these in one oracle. It's difficult to find where the parallelism got adjusted.

Reply:
The number of files isn't known here, so that is not a change we can do easily. It's in the datasource contract that the parallelism is just a "hint" and the actual datasource can decide a lower amount of parallelism if not feasible.

Reply:
Can the parallelism be tuned inside Reader? Like outside Reader, it's still the requested parallelism from user, and then there is a single place to tune/finalize the number inside Reader when all information is ready. It seems no use of parallelism outside Reader?

Reply:
It's currently tuned based on the number of files. Other readers might be reading from a database / REST API and may have other constraints, such as number of connections, etc., that cannot be guessed from outside the reader.